### PR TITLE
[SW-2535] Fix Handling of Flow UI SSL Configuration

### PR DESCRIPTION
**Pull request by mn-mikke**
The state after this PR:
- Sparkling Water Flow UI configuration options just affect the following connections (**SW connections**)
   -  FLOW UI requests from Spark driver to a H2O leader node
   -  data exchange between Spark executors and H2O nodes (Dataframe from/to H2O frame conversions)
- When `spark.ext.h2o.verify_ssl_certificates` is set to `false`, won't verify any ssl certificate on **SW connections**
- When `spark.ext.h2o.auto.flow.ssl` is set to `true`, SW generates self-signed certificate and set it to `spark.ext.h2o.jks` a distributes across the cluster. This certificate is put into a newly created trustore, which used for verification only of **SW connections**.  
- **SW connections** don't verify hostname since we communicated with multiple servers and the certificate is just one.

### CODE REVIEW

File: core/src/main/scala/ai/h2o/sparkling/backend/utils/RestCommunication.scala [hunk 1]
```diff
@@ -227,6 +231,59 @@ trait RestCommunication extends Logging with RestEncodingUtils {
...
+    }
+  }
+
+  private def disableCertificateVerification(connection: HttpsURLConnection): Unit = {
+    val allCertificatesTrustedManager = new X509TrustManager() {
+      def getAcceptedIssuers: Array[X509Certificate] = null
+      def checkClientTrusted(x509Certificates: Array[X509Certificate], s: String): Unit = {}
+      def checkServerTrusted(x509Certificates: Array[X509Certificate], s: String): Unit = {}
+    }
+    val sslContext = SSLContext.getInstance("TLSv1.2")
```
[the remark this answers is not in the document]
Comment (mn-mikke) [reply core/src/main/scala/ai/h2o/sparkling/backend/utils/RestCommunication.scala:261]: Yes, will change. Thanks!

File: core/src/main/scala/ai/h2o/sparkling/backend/utils/RestCommunication.scala [hunk 2]
```diff
@@ -227,6 +231,59 @@ trait RestCommunication extends Logging with RestEncodingUtils {
...
+      val secureConnection = connection.asInstanceOf[HttpsURLConnection]
+      if (!conf.verifySslCertificates) {
+        disableCertificateVerification(secureConnection)
+      } else if (conf.autoFlowSsl) {
+        setSelfSignedCertificateVerification(secureConnection, conf)
+      }
+
+      // Enable all hosts since the certificate located in the distributed jks file could be created
+      // on a different server.
+      secureConnection.setHostnameVerifier(new HostnameVerifier {
```
Comment (michalkurka) [inline core/src/main/scala/ai/h2o/sparkling/backend/utils/RestCommunication.scala:246]: Side note: H2O on Hadoop (h2odriver) has an option to use a per-host JKS entry, via the hostnameAsJksAlias option.

Basically, you can have a JKS that has certificates for each server and they will be looked-up using the hostname as an alias.

**Review by michalkurka — COMMENTED** [review]
I think we need to change the handling of hostname verification or document it clearly.

An argument can be made that we are connecting to a known host obtained from a source we can trust (Spark) and we are thus de facto not compromising the security in any way.

However, we need to be upfront about doing that and possibly provide an option to disable the behavior - users might rather want to chose to issue a wildcard certificate (for example).

File: core/src/main/scala/ai/h2o/sparkling/backend/utils/RestCommunication.scala [hunk 3]
```diff
@@ -227,6 +231,59 @@ trait RestCommunication extends Logging with RestEncodingUtils {
...
+        override def verify(s: String, sslSession: SSLSession): Boolean = true
```
[the remark this answers is not in the document]
Comment (mn-mikke) [reply core/src/main/scala/ai/h2o/sparkling/backend/utils/RestCommunication.scala]: Done, thanks for this comment.

**Review by michalkurka — APPROVED** [review]
LGTM, thanks for making the changes @mn-mikke

**Review by michalkurka — APPROVED** [review]
Thank you @mn-mikke, great to see getting rid off setting system properties at runtime - that can likely cause issues.

**Pull request merged**